### ISSUE TITLE
Big increase of Loss when ReduceLROnPlateau triggers

### ISSUE THREAD

**Issue by johngrabner** [arthurflor23/handwritten-text-recognition#4]
Again, your code is exquisite.

Have you seen a big degradation in loss/val_loss when ReduceLROnPlateau triggers?
In the case of my data, the loss/val_loss is 8.8/5.8 then ReduceLROnPlateau triggers and loss changes to 28/55.

The model is the default flor.  I only put in my data.

```
Epoch 00079: val_loss did not improve from 5.42500
2991/2991 [==============================] - 306s 102ms/step - loss: 8.8125 - val_loss: 5.8467
Epoch 80/1000
2990/2991 [============================>.] - ETA: 0s - loss: 28.50672019-12-29 23:31:56.626872: W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Cancelled: Operation was cancelled

Epoch 00080: val_loss did not improve from 5.42500

Epoch 00080: ReduceLROnPlateau reducing learning rate to 0.00020000000949949026.
2991/2991 [==============================] - 306s 102ms/step - loss: 28.5116 - val_loss: 55.6891
Epoch 81/1000
2990/2991 [============================>.] - ETA: 0s - loss: 45.93922019-12-29 23:37:02.238036: W tensorflow/core/kernels/data/generator_dataset_op.cc:103] Error occurred when finalizing GeneratorDataset iterator: Cancelled: Operation was cancelled

Epoch 00081: val_loss did not improve from 5.42500
2991/2991 [==============================] - 306s 102ms/step - loss: 45.9394 - val_loss: 57.7789
Epoch 82/1000

```

**Comment by johngrabner** [comment]
One more thing, the train losses are lower (5 vs 9) with my data, but the metrics are much worst (word error rate 71% vs 225).

Is there an easy way to dump the transformation to see if the image import worked as expected?

### python main.py --source=grabner --train
                Total train images:      47852
                Total validation images: 8878
                Batch:                   16

                Total time:              397.87 min
                Time per epoch:          4.68 min
                Time per item:           0.00495068 sec

                Total epochs:            85
                Best epoch               65

                Training loss:           3.94768072
                Validation loss:         5.42500214

### python main.py --source=bentham --train

            Total train images:      8711
            Total validation images: 1354
            Batch:                   16

            Total time:              138.53 min
            Time per epoch:          0.84 min
            Time per item:           0.00500474 sec

            Total epochs:            165
            Best epoch               145

            Training loss:           4.93504407
            Validation loss:         9.03654562

###  python main.py --source=bentham --test --arch=flor

                Total test images:    815
                Total time:           0.44 min
                Time per item:        0.03226806 sec

                Metrics:
                Character Error Rate: 0.07796370
                Word Error Rate:      0.22135935
                Sequence Error Rate:  0.77177914


###  python main.py --source=grabner --test 

                Total test images:    9329
                Total time:           2.36 min
                Time per item:        0.01519688 sec

                Metrics:
                Character Error Rate: 0.28574901
                Word Error Rate:      0.71015114
                Sequence Error Rate:  0.71819059

**Comment by arthurflor23** [comment]
Hi again!
This is weird.. first, the ReduceLROnPlateau trigger make the loss decrease a bit until doesn't improve within early stopping patience (you can also change the callbacks values and adjust to your dataset, for example).

About ``Is there an easy way to dump the transformation to see if the image import worked as expected?``, you can see the preprocessed images with --cv2 parameter.. by default you'll see test partition, but you can change to train/valid partition in main.py file.. this images are the data will into to the model (training/validation/predict)

If necessary, you can change the preprocess function in preproc.py, maybe can help too

**Comment by johngrabner** [comment]
Agree the step in loss is weird. I was not watching the training of betham to observe if it happens with this dataset also.  

I have not looked at the cv2 functionality yet, but will.  

One more question, If I may.  Each of the models has one or more dense layers before the ctc.  For these dense to be trained uniformly across the width, there needs to be uniform distribution of training data across the dense layer.  I do not find code that will horizontally offset the input sample so the end of the dense sees as much of variations as the beginning. Without this, the end of the dense will see only opportunities on lines that fill the full 1024.  

Does this make sense?

**Comment by arthurflor23** [comment]
Hi!
I don't know if I understand well, but the inputs of the model are vertical images, matrices with 1024x128 (HxW), where in dense output will be 128xChalistSize (HxW), that is, 128 chars (sentence length) with X possibilities each one. However, you can try use TimeDistributed layer (I used, but nothing changed indeed)

**Comment by johngrabner** [comment]
Thank you for indulging me.

I'm no expert.  I managed to get the Keras example (https://keras.io/examples/image_ocr/) converted to use my sample data a while back.  Also swapped out a couple of models at random.  The Keras example code has a comment:

>         # teach the RNN translational invariance by
>         # fitting text box randomly on canvas, with some room to rotate

I assume this was because the dense layer between the cnn and rnn learns based on the training samples the lands on the specific neurons in the dense layer, while the learning in the cnn and rnn are invariant to the position of based on their nature.

I will do some experiments to see it the above understand holds. 
Fighting issues with my second GPU install and tf 2.0 hickup, so may take a while.

Your 

> def ctc_loss_lambda_func(y_true, y_pred):
>         """Function for computing the CTC loss"""
> 
>         if len(y_true.shape) > 2:
>             y_true = tf.squeeze(y_true)
> 
>         input_length = tf.math.reduce_sum(y_pred, axis=-1, keepdims=False)
>         input_length = tf.math.reduce_sum(input_length, axis=-1, keepdims=True)
>         label_length = tf.math.count_nonzero(y_true, axis=-1, keepdims=True, dtype="int64")
> 
>         loss = K.ctc_batch_cost(y_true, y_pred, input_length, label_length)
>         loss = tf.reduce_mean(loss)
> 
>         return loss

is completely different from the approach the Keras example does it.  

```
    labels = Input(name='the_labels',
                   shape=[img_gen.absolute_max_string_len], dtype='float32')
    input_length = Input(name='input_length', shape=[1], dtype='int64')
    label_length = Input(name='label_length', shape=[1], dtype='int64')
    # Keras doesn't currently support loss funcs with extra parameters
    # so CTC loss is implemented in a lambda layer
    loss_out = Lambda(
        ctc_lambda_func, output_shape=(1,),
        name='ctc')([y_pred, labels, input_length, label_length])
```

Would you be willing to explain how the input_length and especially the label_length is calculated? In Keras example, they feed these in as inputs to the model and move the ctc to the end of the model to use these parameters.  Yours is much more direct, but I don't understand how your code works.

Thank you in advance

**Comment by johngrabner** [comment]
I think I figured it out, is the correct explanation?

```
def ctc_loss_lambda_func(y_true, y_pred):
        """Function for computing the CTC loss"""

        if len(y_true.shape) > 2:
            y_true = tf.squeeze(y_true)

        # y_pred.shape = (batch_size, string_length, alphabet_size_1_hot_encoded)
        # output of every model is softmax
        # so sum across alphabet_size_1_hot_encoded give 1
        #               string_length give string length
        input_length = tf.math.reduce_sum(y_pred, axis=-1, keepdims=False)
        input_length = tf.math.reduce_sum(input_length, axis=-1, keepdims=True)

        # y_true strings are padded with 0
        # so sum of non-zero gives number of characters in this string
        label_length = tf.math.count_nonzero(y_true, axis=-1, keepdims=True, dtype="int64")

        loss = K.ctc_batch_cost(y_true, y_pred, input_length, label_length)

        # average loss across all entries in the batch
        loss = tf.reduce_mean(loss)

        return loss

```

**Comment by arthurflor23** [comment]
Hi @johngrabner !
Indeed, I used/inspired from this [code](https://github.com/ysoullard/CTCModel) and I done with input_length too (some commits ago).. and using a lambda layer, you have other difficulties, such as use 2 models (loss and predict models), save/load models in json.. But after that I understand how the code works, I saw that was possible make more automatic using only one model (loss) and decoding the output separately..

And about the explanation, it's all correct! I'll put this comments in next commit, if you don't mind

**Comment by johngrabner** [comment]
Please feel free to use my comments, I am taking a lot more from you.

Q1: Do you know of any good explanations for TimeDistributed?

**Comment by arthurflor23** [comment]
Not necessarily.. in my tests it didn't make much difference, so I removed it..
After that, I think is that it may not be interesting for the model, since its task is to transcribe text from an image, that is, the "symbols" in the image doesn't need have a relationship between them, unlike a Machine Translation, where the relationship between indices is "stronger".. I may be wrong too

**Comment by johngrabner** [comment]
Thank you for publicizing your code. I have spent the last week dissecting it for the purpose of learning. My key takeaways:
1- The ctc_loss solution is much cleaner than the [Keras OCR example].
(https://keras.io/examples/image_ocr/)
2- The multiprocessing in preprocess_partitions greatly improves performance.  
3- The image quality from preproc.augmentation is better than a simple histogram equalization.
     It does not handle color with opacity, but none of the public data has opacity in the line data.
4- The listing of all public data sets is a great service.
5- The listing of key ocr models is also a very nice thing for you to have done.
